src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+                ++i;
+            } else if (i + 2 < smiles.size() &&
+                       chirality_tags.count(smiles.substr(i + 1, 2)) != 0) {
+                // Set the type of the chirality
+                chirality_type += " " + smiles.substr(i + 1, 2);
+                i += 2;
+                chirality_type += std::to_string(read_number(smiles, ++i));
+            }
+            new_atom.set("chirality", chirality_type);
+            break;
+        default:
+            warning("SMI Reader", "unknown property code: '{}'", smiles[i]);
+            break;
+        }
+        ++i;
+    }
+}
+
+void SMIFormat::read(Frame& frame) {
+    // Initialize all the reading variables
+    branch_point_ = std::stack<size_t>(); 
+    rings_ids_.clear();
+    mol_vector_.clear();
+    current_atom_ = 0;
+    previous_atom_ = 0;
+    current_bond_order_ = Bond::SINGLE;
+    first_atom_ = true;
+
+    Topology topol;
+    size_t groupid = 1;
+    mol_vector_.push_back(Residue("group " + std::to_string(groupid)));
+
+    auto smiles = trim(file_->readline());
+
+    auto check_ring =
+    [&](size_t ring_id) {
+        auto ring_lookup = rings_ids_.find(ring_id);
+
+        if (ring_lookup == rings_ids_.end()) {
+            rings_ids_.insert({ring_id, {previous_atom_, current_bond_order_}});
+            current_bond_order_ = Bond::SINGLE;
+            return;
+        }
+
+        // Deviation from the standard, technically bond orders need to be equal,
+        // but we will accept the stored order if the current order is single.
+        // This is common practice
+        topol.add_bond(previous_atom_,
+            ring_lookup->second.first,
+            current_bond_order_ == Bond::SINGLE?
+                ring_lookup->second.second :
+                current_bond_order_
+        );
+        rings_ids_.erase(ring_lookup);
+
+        current_bond_order_ = Bond::SINGLE;
+    };
+
+    size_t i;
+    for (i = 0; i < smiles.size(); ++i) {
+
+        if (std::isblank(smiles[i]) != 0) {
+            break;
+        }
+
+        if (smiles[i] == '[') {
+            // TODO: Don't pass the full string, instead pass an iterator
+            // or maybe a substring till the closing ']'
+            process_property_list_(topol, smiles, ++i);
+            continue;
+        }
+
+        // We are not in a property list!
+        // Therefore, if something is lowercase, then it must be
+        // a single character element in the organic subset.
+        if (std::islower(smiles[i]) != 0) {
+            if (aromatic_organic.count(smiles[i]) == 0) {
+                throw format_error("SMI Reader", "aromatic element not found: '{}'", smiles[i]);
+            }
+            add_atom(topol, smiles.substr(i, 1)).set("is_aromatic", true);
+            continue;
+        }
+
+        // We are not in a property list!
+        // Therefore, aliphatic atoms can be written without brackets.
+        // Br and Cl are tricky as, but the rest are single character
```

Comment:
Should we account for `At` too as a two letter halogen element?

Reply:
Not according to the standard, but I break from the standard in many places anyway!